CORDA-696: Create separate transaction types for contract upgrade transactions

adagys opened the pull request:
Removed the ContractUpgrade command and created new restricted ContractUpgrade(Wire/Ledger)Transaction types. @mikehearn I'm still not entirety convinced that this is the right approach :)

Modified UpgradedContract to include a contract constraint for the legacy contract.

Created a new vault update type for contract upgrade related changes.

Note that this PR doesn't address the encumbrances issue properly.

kchalkias commented on core/src/main/kotlin/net/corda/core/transactions/ContractUpgradeTransactions.kt:
```diff
@@ -0,0 +1,127 @@
+        override val notary: Party,
+        val legacyContractAttachment: SecureHash,
+        val upgradedContractAttachment: SecureHash,
+        val privacySalt: PrivacySalt = PrivacySalt()
```
Should we use privacySalt in the MTree construction? Now, it seems to be a redundant property.

adagys replied:
Good catch! forgot to actually include it

kchalkias commented on core/src/main/kotlin/net/corda/core/transactions/ContractUpgradeTransactions.kt:
```diff
@@ -0,0 +1,127 @@
+
+/** A special transaction for upgrading the contract of a state. */
+@CordaSerializable
+data class ContractUpgradeWireTransaction(
```
Are we interested in backwards compatibility for the new tx_types?

adagys replied:
I suspect we are

kchalkias replied:
Nice, we should define what is considered backwards compatibility. If a new property is added, do we care if the old client can still construct the MTree?

kchalkias commented on core/src/main/kotlin/net/corda/core/transactions/ContractUpgradeTransactions.kt:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    /** Hash of the component list that are hidden in the [ContractUpgradeFilteredTransaction]. */
+    private val hiddenComponentHash: SecureHash get() = serializedHash(outputs + legacyContractAttachment + upgradedContractAttachment)
```
It would be better and more secure to apply either an HMAC or double hashing (as we do in `WireTransaction`) to protect against length extension attacks. FYI, SHA-256 (our default hash algorithm) is vulnerable against the aforementioned attack.

adagys replied:
Ok, will change to double hashing

rnicoll commented on core/src/test/kotlin/net/corda/core/node/VaultUpdateTests.kt:
```diff
@@ -12,7 +12,8 @@ import kotlin.test.assertFailsWith
 class VaultUpdateTests {
-    val DUMMY_PROGRAM_ID = "net.corda.core.node.VaultUpdateTests.DummyContract"
+    private val DUMMY_PROGRAM_ID = "net.corda.core.node.VaultUpdateTests.DummyContract"
+    private val noUpdate = Vault.Update(emptySet(), emptySet())
```
Not sure why you're adding a new `noUpdate` value?

adagys replied:
I removed the old one from the public API – a "no update" update is an implementation detail and shouldn't really be something that the vault emits

rnicoll commented on core/src/main/kotlin/net/corda/core/internal/ResolveTransactionsFlow.kt:
```diff
@@ -156,13 +158,18 @@ class ResolveTransactionsFlow(private val txHashes: Set<SecureHash>,
-        val wireTransactions = downloads.filterNot { it.isNotaryChangeTransaction() }.map { it.tx }
-        val missingAttachments = wireTransactions.flatMap { wtx ->
-            wtx.attachments.filter { serviceHub.attachments.openAttachment(it) == null }
+        val attachments = downloads.flatMap {
```
Might be a little cleaner as `val attachments = downloads.coreTransaction.flatMap { coreTransaction ->`

adagys replied:
do you mean something like `downloads.map { it.coreTransaction }.flatMap { coreTransaction -> .. }?

rnicoll replied:
I misread the code, but yes lets go with your revised version :)

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/ContractUpgradeTransactions.kt:
```diff
@@ -0,0 +1,128 @@
+        val privacySalt: PrivacySalt,
+        override val sigs: List<TransactionSignature>
+) : FullTransaction(), TransactionWithSignatures {
+    private val legacyContractName: ContractClassName = inputs.first().state.contract
```
OK, so the inputs must have a consistent constraint - that should be in the docs.

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/ContractUpgradeTransactions.kt:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    override fun verify() {
+        val upgradedContract = when (upgradedContractTry) {
```
Why is the logic that loads the contract outside the verify function? That isn't obvious and if really required should have a comment explaining it. Otherwise I'd lose the Try and use a plain old fashioned try/catch.

adagys replied:
As in `LedgerTransaction` the whole `verify()` method is/will be run with blacklisting enabled, so we need to do the contract loading beforehand. Will add a comment

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/ContractUpgradeTransactions.kt:
```diff
@@ -0,0 +1,128 @@
+            "Outputs' contract must be an upgraded version of the inputs' contract"
+        }
+
+        inputs.forEachIndexed { index, stateAndRef ->
```
This is probably better written as `for ((index, stateAndRef) in inputs.indexed()) { .. }` - intellij can auto-convert between forms.

rnicoll reviewed (commented):
I'm unclear why this would be the wrong approach, but a separate transaction type makes sense for notary change transactions? It feels to me that the same structure applies to both.

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/NotaryChangeTransactions.kt:
```diff
@@ -27,7 +27,8 @@ data class NotaryChangeWireTransaction(
      */
     override val outputs: List<TransactionState<ContractState>>
-        get() = emptyList()
+        get() = throw IllegalStateException("NotaryChangeWireTransaction does not contain output states, " +
```
UnsupportedOperationException

adagys replied:
I'm especially uncomfortable about this part

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt:
```diff
@@ -50,17 +50,14 @@ data class SignedTransaction(val txBits: SerializedBytes<CoreTransaction>,
+    val coreTransaction: CoreTransaction get() = cachedTransaction ?: txBits.deserialize().apply { cachedTransaction = this }
 
-    /** Returns the contained [NotaryChangeWireTransaction], or throws if this is a normal transaction. */
-    val notaryChangeTx: NotaryChangeWireTransaction get() = transaction as NotaryChangeWireTransaction
```
This isn't backwards compatible.

adagys replied:
In practice this is (or should be) internal API, I don't think we should maintain it forever just for the sake of it.

rnicoll replied:
I've re-added with an `@Deprecated` annotation in the next version, so in the highly unlikely event someone has compiled against it, they'll have some warning it's going to disappear.

mikehearn commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -118,81 +115,74 @@ class NodeVaultService(private val clock: Clock, private val keyManagementServic
-                        regularTxns.clear()
-                    }
-                }
+            if (batch.isNotEmpty() && tx::class.java != batch.last()::class.java) {
```
not `.javaClass`?

mikehearn commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/Kryo.kt:
```diff
@@ -264,6 +268,29 @@ object NotaryChangeWireTransactionSerializer : Serializer<NotaryChangeWireTransa
 @ThreadSafe
+object ContractUpgradeWireTransactionSerializer : Serializer<ContractUpgradeWireTransaction>() {
+    override fun write(kryo: Kryo, output: Output, obj: ContractUpgradeWireTransaction) {
+        kryo.writeClassAndObject(output, obj.inputs)
```
Does this pass the AMQP CI gate?

adagys replied:
Probably not :(

mikehearn commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -118,81 +115,74 @@ class NodeVaultService(private val clock: Clock, private val keyManagementServic
     override val updates: Observable<Vault.Update<ContractState>>
         get() = mutex.locked { _updatesInDbTx }
 
+    /** Groups adjacent transactions into batches to generate separate net updates per transaction type. */
```
Does this new implementation emit things in the same order as before, or did it change?

adagys replied:
It does, the old approach didn't scale well to more than two types so I had to rewrite it.

mikehearn replied:
Is there a quick way to get the old ordering, accepting the bad scaling, with the new code still present but `if (false)`d out so we can gate it behind targetVersion once that's implemented? Then apps can opt in to the new faster code.

adagys replied:
Sorry I meant it doesn't scale in terms of code structure 😄 The ordering is still the same

mikehearn commented:
I think Andrius didn't like the separate tx approach last time either :) It's me who's pushing for that style.

I quite like how the upgrade code is looking now.  The main areas I'm unsure about and are triggering comments are:

* Compatibility
* The interaction with constraints
* To what extent we should be pushing this mechanism on users vs just using relaxed code signature constraints

It's good that we have a protocol with this level of control, but I suspect in practice - at least in the early days - nobody will be using really immutable contracts. They'll be happy to let third parties patch the code on the fly. But we'll see.

rnicoll commented:
After discussion with @adagys I think the upgrade transaction shouldn't include outputs (as per notary change); this means resolving a transaction chain includes rebuilding the upgraded states. However as verifying the transaction chain would require rebuilding them anyway, this doesn't seem unduly expensive.

mikehearn commented:
Good point.

adagys commented on core/src/main/kotlin/net/corda/core/contracts/Structures.kt:
```diff
@@ -246,6 +246,7 @@ annotation class LegalProseReference(val uri: String)
  */
 interface UpgradedContract<in OldState : ContractState, out NewState : ContractState> : Contract {
     val legacyContract: ContractClassName
+    val legacyContractConstraint: AttachmentConstraint
```
Hmm, true, but this more of a security fix really

rnicoll replied:
I'm inclined that we force them to add the new property; we can't support the old interface in actual transactions, so keeping it for recompiling seems less friendly than saying "Either add this property or remove the interface".

adagys commented on core/src/main/kotlin/net/corda/core/node/services/TransactionVerifierService.kt:
```diff
@@ -14,5 +14,5 @@ interface TransactionVerifierService {
      * @return A future that completes successfully if the transaction verified, or sets an exception the verifier threw.
      */
-    fun verify(transaction: LedgerTransaction): CordaFuture<*>
+    fun verify(transaction: FullTransaction): CordaFuture<*>
```
I thought we don't yet provide ABI stability?

adagys commented on core/src/main/kotlin/net/corda/core/internal/ContractUpgradeUtils.kt:
```diff
@@ -1,21 +1,28 @@
+            val oldState = it.data
+            val newState = contractUpgrade.upgrade(oldState)
+            // TODO: copy across encumbrances
+            TransactionState(newState, upgradedContractClass.name, it.notary, it.encumbrance, upgradedContractConstraint)
```
Not exactly sure what you mean (?)

adagys commented on core/src/main/kotlin/net/corda/core/flows/NotaryFlow.kt:
```diff
@@ -65,12 +68,15 @@ class NotaryFlow {
                     }
-                    session.sendAndReceiveWithRetry(tx)
+
+                    session.sendAndReceiveWithRetry(filteredTransaction)
```
That's a good question, I also suspect it won't be

adagys commented on core/src/main/kotlin/net/corda/core/internal/ResolveTransactionsFlow.kt:
```diff
@@ -156,13 +158,18 @@ class ResolveTransactionsFlow(private val txHashes: Set<SecureHash>,
      * Returns a list of all the dependencies of the given transactions, deepest first i.e. the last downloaded comes
      * first in the returned list and thus doesn't have any unverified dependencies.
      */
+    // TODO: This could be done in parallel with other fetches for extra speed.
```
I pulled it out since it made it seem like it applied to the specific line below

andr3ej commented:
having discussed with @adagys i've added a commit that fixes the circular dep between NodeVaultService and ServiceHub

rnicoll commented:
@mikehearn Okay, I think this is ready for re-review. Aware we have API functions removed, I can put them back with stubs/exception throwing versions, but realistically I think everything removed shouldn't have been public and is of limited/no value to CorDapps :-/

mikehearn commented:
OK, this is on my review queue again, sorry for being slow ...

rnicoll commented:
Quick reminder, Mike

mikehearn commented:
I had not forgotten. I do actually have a todo list, this is currently in 7th place (about the middle).

Clintonio reviewed (changes requested):
Not much to say - the changes to the areas I've worked on seem fine.

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/ContractUpgradeTransactions.kt:
```diff
@@ -0,0 +1,150 @@
+
+    /**
+     * This transaction does not contain any output states, outputs can be obtained by resolving a
+     * [ContractUpgradeLedgerTransaction] and applying the notary modification to inputs.
```
"applying the notary modification" seems like a copy/paste error.

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/ContractUpgradeTransactions.kt:
```diff
@@ -0,0 +1,150 @@
+    /** Resolves input states and contract attachments, and builds a ContractUpgradeLedgerTransaction. */
+    fun resolve(services: ServicesForResolution, sigs: List<TransactionSignature>) = resolve(services, services.attachments, sigs)
+
+    fun resolve(stateLoader: StateLoader, attachments: AttachmentStorage, sigs: List<TransactionSignature>): ContractUpgradeLedgerTransaction {
```
Does this need a kdoc?

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/ContractUpgradeTransactions.kt:
```diff
@@ -0,0 +1,150 @@
+        val resolvedInputs = inputs.map { ref ->
+            stateLoader.loadState(ref).let { StateAndRef(it, ref) }
+        }
+        val legacyContractClassName = resolvedInputs.first().state.contract
```
Do we need to check consistency here?

adagys replied:
The check is done later in the ContractUpgradeLedgerTransaction. I think it's easier to keep all checks in one place

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/ContractUpgradeTransactions.kt:
```diff
@@ -0,0 +1,150 @@
+) : FullTransaction(), TransactionWithSignatures {
+    private val upgradedContract: UpgradedContract<ContractState, *> by lazy {
+        @Suppress("UNCHECKED_CAST")
+        this::class.java.classLoader.loadClass(upgradedContractAttachment.contract).
```
This needs a TODO to load it from the CorDapp classloader (and eventually the sandboxing classloader).

And we should really centralise this code as it's so critical.... can we refactor the existing codepath that creates the contract object so this code can use it?

Also this is lazy, but then there's a parameter below that eagerly resolves it - seems the lazyness is pointless.

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/ContractUpgradeTransactions.kt:
```diff
@@ -0,0 +1,150 @@
+            "All input states must point to the legacy contract"
+        }
+        check(inputs.all { it.state.constraint.isSatisfiedBy(legacyContractAttachment) }) {
+            "Legacy contract constrain does not satisfy the constraint of the input states"
```
typo: constrain->constraint

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/NotaryChangeTransactions.kt:
```diff
@@ -40,7 +41,9 @@ data class NotaryChangeWireTransaction(
      */
     override val id: SecureHash by lazy { serializedHash(inputs + notary + newNotary) }
 
+    @Deprecated("Exists purely for binary compatibility, the more generic version based on StateLoader should be used instead")
```
Yep. Could even use the kotlin `@ReplaceWith` annotation in this case as it's so trivial.

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt:
```diff
@@ -152,71 +155,96 @@ data class SignedTransaction(val txBits: SerializedBytes<CoreTransaction>,
+        resolveNotaryChangeTransaction(services)
+    }
+
+    /** TODO */
```
???

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt:
```diff
@@ -152,71 +155,96 @@ data class SignedTransaction(val txBits: SerializedBytes<CoreTransaction>,
+    }
+
     /**
      * TODO: Verify contract constraints here as well as in LedgerTransaction to ensure that anything being deserialised
```
I wonder if this TODO is still relevant.

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt:
```diff
@@ -152,71 +155,96 @@ data class SignedTransaction(val txBits: SerializedBytes<CoreTransaction>,
-            is NotaryChangeWireTransaction -> resolveNotaryChangeTransaction(services)
+    fun resolveBaseTransaction(services: ServicesForResolution) = resolveBaseTransaction(services, services.attachments)
+
+    fun resolveBaseTransaction(stateLoader: StateLoader, attachments: AttachmentStorage): BaseTransaction {
```
I guess you could use `@JvmOverloads` with a default argument. At any rate, please copy the Kdoc so all public members are documented.

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt:
```diff
@@ -152,71 +155,96 @@ data class SignedTransaction(val txBits: SerializedBytes<CoreTransaction>,
+     */
+    fun resolveContractUpgradeTransaction(services: ServicesForResolution) = resolveContractUpgradeTransaction(services, services.attachments)
+
+    fun resolveContractUpgradeTransaction(stateLoader: StateLoader, attachments: AttachmentStorage): ContractUpgradeLedgerTransaction {
```
kdoc

adagys committed:
Create separate transaction types for contract upgrade transactions.

Fix circular dependency. (#2011)

Add rationale around upgrade transactions

Move contract upgrade transaction resolution logic into internal until it's stabilised.

adagys commented:
@mikehearn this is ready for another look. The API checker still reports 2 issues, but I don't see a way around them

mikehearn commented:
It's reporting one issue, right?

[11:36:27][Step 5/10] -@net.corda.core.DoNotImplement public interface net.corda.core.node.services.TransactionStorage extends net.corda.core.node.StateLoader
[11:36:27][Step 5/10] +@net.corda.core.DoNotImplement public interface net.corda.core.node.services.TransactionStorage

So TransactionStorage no longer extends StateLoader. How comes?

mikehearn commented on core/src/main/kotlin/net/corda/core/node/services/TransactionStorage.kt:
```diff
@@ -14,18 +14,12 @@ import rx.Observable
      */
     fun getTransaction(id: SecureHash): SignedTransaction?
 
-    @Throws(TransactionResolutionException::class)
```
Why does this need to be removed?

mikehearn replied:
Ping? Isn't this an API break from 1.0?

mikehearn replied:
OK, had a chat with Andrius, it's not because this API change is a revert of something that never shipped in 1.0/2.0

adagys commented:
Well `TransactionStorage` has been recently refactored to extend `StateLoader`, which made sense at the time, but with contract upgrade transactions it now needs access to `AttachmentStorage` and later potentially `CordappProvider`, which means we have to wire it up in all constructors of implementations (there are multiple), and that gets messy. Maybe @andr3ej has any suggestions?

adagys commented:
And the other issue is 
`+  @org.jetbrains.annotations.NotNull public abstract net.corda.core.contracts.AttachmentConstraint getLegacyContractConstraint()`

gendal commented:
Any docs?

gendal commented:
Question: if this did not go in to 3.0, what would be the consequence? ie would subsequently implementing this break wire/api/abi compat in some way?

adagys commented:
It would break backwards compatibility if someone were to start using the contract upgrade mechanism – which doesn't fully work at the moment. We _may_ be able to maintain compatibility, but that would be extremely painful.

andr3ej commented:
StateLoader stuff looks fine

mikehearn commented on core/src/main/kotlin/net/corda/core/contracts/Structures.kt:
```diff
@@ -254,7 +254,14 @@ annotation class LegalProseReference(val uri: String)
+    /**
+     * A validator for the legacy (pre-upgrade) contract attachments on the transaction.
+     */
+    val legacyContractConstraint: AttachmentConstraint
```
Hm, so, let's think about this a bit. That would be an API break, albeit an understandable one given the half baked nature of what shipped previously here. Can we provide a non-stupid default here, that at least matches the prior behaviour (even if slightly broken)? I'm thinking specifically of converting this interface to Java and then making this getter have a default implementation that returns AlwaysAcceptAttachmentConstraint. WDYT?

mikehearn replied:
OK, had a catchup with Andrius about this.

Here's what we agreed -

1. Move this interface into a Java file so we can use Java 8 default interface methods.
2. Make getLegacyContract() automatic: it should just use reflection to figure out the name of the current class, which is a good default guess
3. Make getLegacyContractConstraint() return WhitelistedByNetworkParametersAttachmentConstraint, which is being added in PR #2474 by Tudor. This constraint is the only one we can provide without any context that isn't a security hole, as AcceptAllAttachmentConstraint would be, and is a non-terrible default that lets us dig developers out of a hole if they make one for themselves. In future versions of the platform we can use targetVersion of the calling app to change the default to a code signing constraint, initialised to the same signer as the calling app. So in all cases there's a sane set of defaults for these methods.

That requires Andrius to wait for Tudor to go into master and then 3.0, and then Andrius follows close behind, but that's OK because Andrius still needs to do some end-user testing and documentation writing. So there's a bit of time left on this PR before it can go in anyway.

adagys replied:
So I tried moving the interface to Java, but that now breaks existing Kotlin implementations that override the `legacyContract` with a property, you need to change
`override val legacyContract = ...` into `override fun getLegacyContract(): String { ... }`

adagys replied:
To your second point, `getLegacyContract()` should point to the class it's upgrading, and not the current one (which we assume will have a different name), so making it automatic is tricky

mikehearn commented on core/src/main/kotlin/net/corda/core/internal/ContractUpgradeUtils.kt:
```diff
@@ -1,21 +1,31 @@
+    ): ContractUpgradeWireTransaction {
+        require(stateAndRef.state.encumbrance == null) { "Cannot upgrade an encumbered state" }
+
+        val legacyContractAttachmentId = services.cordappProvider.getContractAttachmentID(stateAndRef.state.contract)!!
```
Are we guaranteed to never get a null here? If not can we throw a better exception than NPE?

mikehearn commented on core/src/test/kotlin/net/corda/core/flows/ContractUpgradeFlowTest.kt:
```diff
@@ -226,9 +214,14 @@ class ContractUpgradeFlowTest {
     }
 
     class CashV2 : UpgradedContract<Cash.State, CashV2.State> {
+        override val legacyContractConstraint = object : AttachmentConstraint {
```
This is the same as `get() = AlwaysAcceptAttachmentConstraint`, but if we make that the default impl, this can go away.

mikehearn reviewed (changes requested):
Looking really close now. Have you tried using it with a sample app, like an end user would? Like running a network with demobench and seeing the upgrade go through.

adagys committed:
Throw a better exception when contract attachment not found

adagys committed:
Fix attachment storage init

adagys committed:
Fix an issue where upgraded state retained original state's contract reference

adagys commented:
One other assumption the new contract upgrade transactions make is that the contract attachment will also contain the state code. Is that safe to assume?

adagys commented:
@tudor-malene see my above comment

tudor-malene commented:
Q: One other assumption the new contract upgrade transactions make is that the contract attachment will also contain the state code. Is that safe to assume?

A: For 3.0, definitely, CorDapps need to be fat jars, otherwise the whitelist/hash based security we have in place is useless

adagys commented:
Right, but do we need to add support for potential other attachments containing states only? In the future

adagys committed:
Default legacy contract constraint to always accepting - needs to be changed to whitelist constraint before merging

tudor-malene commented:
I think @mikehearn said that we'll keep the fatjars until we upgrade to java 9, which has built-in support for dependencies.

adagys committed:
Introduce a new upgraded contract interface that allows specifying the legacy constraint.

adagys committed:
Update api

adagys committed:
Make sure contracts implementing the new interface get loaded

adagys commented:
Continuing on #2589
